### DISCUSSION TITLE
Issues in running mHM

### DISCUSSION THREAD

**Discussion by Mohammedhhu** [mhm-ufz/mHM#111] [General]
Dear mHM expert 

Recently, I have set up and prepared all the input of mHM. I am running mHM across the whole Nile River Basin (3,176,541 km2). I got this attached error, and I think it is in regard to my Latlon file. 
L0 = 0.00902777785
L1 = 0.25
L11 = 0.5
I have attached the error and the information of latlon.nc for your reference. 
I am looking for your suggestion and recommendation to solve this issues 

![image](https://github.com/user-attachments/assets/930c388a-853f-45d2-aa2c-ebf7c997c2f2)
![image](https://github.com/user-attachments/assets/77825e57-1e38-41e6-ad20-3c60c03e1799)

[mhm.zip](https://github.com/user-attachments/files/17935376/mhm.zip)
[header_1.txt](https://github.com/user-attachments/files/17935346/header_1.txt)
[header_2.txt](https://github.com/user-attachments/files/17935348/header_2.txt)
[header_3.txt](https://github.com/user-attachments/files/17935349/header_3.txt)

**Comment by ViniciusBMoreira** [comment]
I believe that all layers need to be multiples of each other and have the same extensions. L0, for example, doesn't follow this rule

**Reply by Mohammedhhu** [reply]
I have checked all the morphological data regarding to the grid scale and extent. 
After preparing new morphological files to the same extent as the meteorological ones, I got a new error. 
Indeed, my soil_classification.txt has no value such as -9999.

what dose it mean?
![image](https://github.com/user-attachments/assets/c7e40f50-684e-41c0-911c-06768198d984)

**Reply by StephanThober** [reply]
Dear Mohammedhhu,

the mask in mHM is derived from the dem file. This error indicates that in your soil_class.asc you have a missing value at a grid cell where the dem has a value. Did you check that the mask in the dem.asc and soil_class.asc are the same?

Best,
Stephan

**Reply by Mohammedhhu** [reply]
Dear StephanThober 

Thank you so much for your observations; yes, there are some pixels with missing data in soil and geology maps while they have values in dem map. I have solved this by filling in the missing data. 

Best regards, 
Mohammed

**Comment by Mohammedhhu** [comment]
Dear all

Could you please help diagnose the cause of this segmentation fault? Are there specific checks or debugging steps you recommend to resolve this issue?

Thank you for your time and support. Please let me know if you need additional details or files.


![image](https://github.com/user-attachments/assets/ed586db4-a192-44ad-b003-ce6f9d2a80a3)

**Reply by pallavhydro** [reply]
Hi @Mohammedhhu 


The fastest way to workout the error in segfault is to run mhm in debug mode. 

#### Option 1
If you compiled mhm in your own conda environment, you can make a debug version of the executable by running the following command at your mHM repository folder
```
source CI-scripts/compile_debug
```
Run the debug executable and that should provide more insights on the issue during runtime.

#### Option 2
If you are using the one liner installation of mHM with `conda install mhm`, you might not be able to get the debug version using this method (@StephanThober @MuellerSeb please confirm). You would then have to make a conda environment based compilation of mHM as given in step-by-step procedure in the mHM documentation [here](https://mhm.pages.ufz.de/mhm/stable/md_doc_2_i_n_s_t_a_l_l.html#autotoc_md3).


Let us know if you could compile the debug executable and find any clues.


Cheers
Pallav

**Reply by Mohammedhhu** [reply]
Hi Pallavhydro
Thank you so much for the information and steps that you provided. I have flow the second option and I create conda environment with full compilation. Indeed the test examples coming with mhm file running well. But I have same error even after this step.

**Reply by pallavhydro** [reply]
Hi @Mohammedhhu 

Thanks for reporting back.
So the debug version of mhm (mhm_debug) also doesn't provide any different runtime error message? Could you please share the error this time.

P

**Reply by Mohammedhhu** [reply]
![image](https://github.com/user-attachments/assets/23eb609d-fd5f-49d0-9e4c-8b3b26c2275e)

**Reply by pallavhydro** [reply]
Interesting. 

I am a bit perplexed. Usually, when we compile and run the debug version, the runtime error (backtrace) should show specifics on the modules and line numbers.

Are you sure you "successfully" compiled the debug version by:
`source CI-scripts/compile_debug`

And then run the corresponding and newly compiled `mhm_debug` file as 
`./mhm_debug`

Important: Please make sure that the executable corresponds to the compilation. You can do that my checking the time stamp of the executable which should correspond to the time of compilation. It is very common, even amongst us developers, that we overlook the completion of compilation, and wrongly use the previously compiled executable. I hope that is not your case (??).

KR
Pallav

**Reply by Mohammedhhu** [reply]
Yes, I have successfully compiled the debug version, but when I ran, I just used ./mhm without using the ./mhm_debug.

From this, I think the error from some input data, I will check them accordingly. 

Any, recommendation ....

![image](https://github.com/user-attachments/assets/9ca07718-fd96-48ab-be05-ab1bc6e94edf)

**Comment by pallavhydro** [comment]
Great to see you got more meaningful runtime error message using the debug executable. 

As per the message, you have at least one occurrence of -9998 in one or more of your input soil maps. Finding the -9998/s and replacing it/them by -9999 (standard noData value) should help.

Meanwhile, also please share the soil classification file (or screenshot of) you are using for your setup for a quick check.


Cheers
Pallav

**Reply by Mohammedhhu** [reply]
Yeah, I will correct this error and replace the -9998 with -9999. 
This is my soil data file for your reference. 


Thank you so much for your help. I do appreciate.

Best regards,
Mohammed  
[Soil.zip](https://github.com/user-attachments/files/18064447/Soil.zip)

**Reply by pallavhydro** [reply]
Hi @Mohammedhhu 

1. Were you able to go past the last error message?
2. How many soil layers do you have in your model? You shared only one file.
3. If you want, I can quickly check your mhm.nml for any possible source of issues.


P

**Reply by Mohammedhhu** [reply]
Hi @pallavhydro 

I have checked the soil class to ensure that is not content any values such as -9998. While I have got the same error. 
Regarding the soil layer, I have just used the soil source from Harmonized World Soil. 
I am attaching the mhm.nml to you here for your reference

[mhm.zip](https://github.com/user-attachments/files/18106310/mhm.zip)

**Reply by pallavhydro** [reply]
Hi @Mohammedhhu 


I checked the mHM source code for the issue. You need to revise the `soil_classdefinition.txt` file for the following points -

1. Revise the `nSoil_Types` to the number of soil types in the file. It is currently the largest ID.
2. The `MU_GLOBAL` needs to be in sequential order of 1, 2, 3, ... currently it is 2, 37, 39, 47, ... So you have to rename 2 to 1, 37 to 2, and so on.
4. You have soil IDs and horizons with multiple entries. Please only keep the correct entries. E.g. `16339` has two entries each for horizon 1 and 2.

Attached is the screenshot comparing `your file (left)` with the `reference test_domain file (right)`: 

![Screenshot 2024-12-12 at 10 40 12](https://github.com/user-attachments/assets/ad550795-281b-477e-8c09-ac83dc349a6a)


Hope this helps
Pallav

**Reply by Mohammedhhu** [reply]
Hi @pallavhydro 

I hope this message finds you well.

I followed your suggestion to renumber the MU_GLOBAL with 1, 2, 3 instead of using the numbers derived from the soil file. However, the same error persists. I am uncertain why this issue continues, despite thoroughly checking multiple times.

To ensure I haven’t missed any critical steps or made an error, I’ve outlined below the process I followed:

    I downloaded the HWSD and added the hwsd.bil file to ArcMap for resampling and masking using my study area shapefile.
    The masked raster file was converted to ASCII format.
    The soil_classdefinition.txt file was generated from the masked raster soil file and the soil database using an R script.

I have attached the masked raster soil map and the R code for your reference. I would greatly appreciate your advice and suggestions to help resolve this issue.

Looking forward to your insights.

Best regards,
Mohammed
[Soil_NRB.zip](https://github.com/user-attachments/files/18196547/Soil_NRB.zip)


![image](https://github.com/user-attachments/assets/eec18880-50b5-44bd-a937-f7318d9734a3)

**Reply by pallavhydro** [reply]
Hi @Mohammedhhu 

The change of the IDs to 1,2,3,... in the definition/ look up file (`soil_classdefinition.txt`) means your soil map (`soil_class.asc`) should also be adapted accordingly. Only then mhm would correctly link the spatial information and the look up table information. Since the error message is still looking for an ID based on previous system (16487) I'm assuming the adaptation of the soil IDs in soil maps is still pending, right?

Pallav

**Reply by Mohammedhhu** [reply]
Hi @pallavhydro 

I have changed the soil_class.asc, and my model runs well. Thank you so much for your help, I do appreciate your support.
Thank you so much to you guys @StephanThober @BryanSouzaM

**Reply by pallavhydro** [reply]
Hi @Mohammedhhu 

I'm happy your setup is running.

To bring a closure to this discussion, I'd appreciate if you set the discussion as answered.

Cheers
Pallav
